quinn/src/delay_queue.rs

```diff
@@ -0,0 +1,442 @@
+use std::{fmt, ops::RangeInclusive};
+
+use slab::Slab;
+
+/// Stores values to be yielded at specific times in the future
+#[derive(Debug)]
+pub struct DelayQueue<T> {
+    /// Definitions of each active timer
+    timers: Slab<TimerState<T>>,
+
+    /// Hierarchical timer wheel
+    ///
+    /// Timers are bucketed into slots organized into levels. Level n spans 2^(`LOG_2_SLOTS` *
+    /// (n+1)) ticks, and each of its slots corresponds to a span of 2^(`LOG_2_SLOTS` * n). The
+    /// expiry time of the first slot in each level is always a multiple of the level's size. When
+    /// the expiry time of a slot in a level above 0 is reached, timers stored in that slot are
+    /// redistributed to lower levels. When the last slot in a level expires, the level advances to
+    /// cover the next span of ticks.
+    ///
+    /// Timers always use the lowest possible level. Every level above 0 has an empty slot,
+    /// corresponding exactly to the span of the previous level.
+    ///
+    /// The set of timers in a slot is an intrusive doubly-linked list, allowing constant-time
+    /// insertion and removal.
+    levels: [Level; LEVELS],
+
+    /// Earliest point at which a timer may be pending
+    ///
+    /// Each `LOG_2_SLOTS` bits of this are a cursor into the associated level, in order of
+    /// ascending significance.
+    next_tick: u64,
```

Comment:
I assume the queue assumes no specific time unit (like `ms`) and just supports any, as long as the user is consistent using the same?

Reply:
Right. Added some discussion to the type's documentation.